ccl/changefeedccl: TestChangefeedDescription failed under stress

SHA: https://github.com/cockroachdb/cockroach/commits/71681f60031c101f17339236e2ba75f7a684d1a1

Parameters:
```
TAGS=
GOFLAGS=-race
```

To repro, try:

```
# Don't forget to check out a clean suitable branch and experiment with the
# stress invocation until the desired results present themselves. For example,
# using stress instead of stressrace and passing the '-p' stressflag which
# controls concurrency.
./scripts/gceworker.sh start && ./scripts/gceworker.sh mosh
cd ~/go/src/github.com/cockroachdb/cockroach && \
stdbuf -oL -eL \
make stressrace TESTS=TestChangefeedDescription PKG=github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl TESTTIMEOUT=5m STRESSFLAGS='-maxtime 20m -timeout 10m' 2>&1 | tee /tmp/stress.log
```

Failed test: https://teamcity.cockroachdb.com/viewLog.html?buildId=1155512&tab=buildLog

```
  github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTask()
      /go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:320 +0x14d
  github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl.(*changeAggregator).Start()
      /go/src/github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl/changefeed_processors.go:180 +0x14cb
  github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl.(*changeFrontier).Start()
      /go/src/github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl/changefeed_processors.go:403 +0x96
  github.com/cockroachdb/cockroach/pkg/sql/distsqlrun.(*ProcessorBase).Run()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/distsqlrun/processors.go:800 +0x8a
  github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl.(*changeFrontier).Run()
      <autogenerated>:1 +0x57
  github.com/cockroachdb/cockroach/pkg/sql/distsqlrun.(*Flow).Run()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/distsqlrun/flow.go:623 +0x144
  github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).Run()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:252 +0xd19
  github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl.distChangefeedFlow()
      /go/src/github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl/changefeed_dist.go:191 +0x1292
  github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl.(*changefeedResumer).Resume()
      /go/src/github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl/changefeed_stmt.go:420 +0x5be
  github.com/cockroachdb/cockroach/pkg/jobs.(*Registry).resume.func1()
      /go/src/github.com/cockroachdb/cockroach/pkg/jobs/registry.go:542 +0x3ff
  github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTask.func1()
      /go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:325 +0xf3
==================
W190228 08:00:32.941702 35986 storage/intentresolver/intent_resolver.go:829  failed to gc transaction record: could not GC completed transaction anchored at /Local/Range/Table/53/RangeDescriptor: node unavailable; try another peer
I190228 08:00:32.942048 35540 util/stop/stopper.go:546  quiescing; tasks left:
2      [async] storage.IntentResolver: cleanup txn records
1      [async] transport racer
1      [async] job-430146954276929537
1      [async] intent_resolver_ir_batcher
1      [async] intent_resolver_gc_batcher
1      [async] closedts-rangefeed-subscriber
1      [async] changefeed-poller
I190228 08:00:32.943542 35540 util/stop/stopper.go:546  quiescing; tasks left:
1      [async] transport racer
1      [async] storage.IntentResolver: cleanup txn records
1      [async] job-430146954276929537
1      [async] intent_resolver_ir_batcher
1      [async] closedts-rangefeed-subscriber
1      [async] changefeed-poller
I190228 08:00:32.943785 35540 util/stop/stopper.go:546  quiescing; tasks left:
1      [async] transport racer
1      [async] job-430146954276929537
1      [async] intent_resolver_ir_batcher
1      [async] closedts-rangefeed-subscriber
1      [async] changefeed-poller
I190228 08:00:32.945035 35540 util/stop/stopper.go:546  quiescing; tasks left:
1      [async] transport racer
1      [async] job-430146954276929537
1      [async] closedts-rangefeed-subscriber
1      [async] changefeed-poller
I190228 08:00:32.946114 35540 util/stop/stopper.go:546  quiescing; tasks left:
1      [async] transport racer
1      [async] job-430146954276929537
1      [async] closedts-rangefeed-subscriber
I190228 08:00:32.948544 35854 ccl/changefeedccl/changefeed_stmt.go:444  [n1] CHANGEFEED job 430146954276929537 returning with error: open /tmp/TestChangefeedDescription_enterprise036353357/ca: no such file or directory
I190228 08:00:32.952119 35540 util/stop/stopper.go:546  quiescing; tasks left:
1      [async] transport racer
1      [async] job-430146954276929537
I190228 08:00:32.952356 35540 util/stop/stopper.go:546  quiescing; tasks left:
1      [async] transport racer
I190228 08:00:33.079375 35787 kv/transport_race.go:113  transport race promotion: ran 13 iterations on up to 511 requests
    --- FAIL: TestChangefeedDescription/enterprise (121.96s)
        testing.go:771: race detected during execution of test

```

SHA: https://github.com/cockroachdb/cockroach/commits/6e0d7081547ee76e52c5f798b78ff739feec2d06

Parameters:
```
TAGS=
GOFLAGS=-race
```

To repro, try:

```
# Don't forget to check out a clean suitable branch and experiment with the
# stress invocation until the desired results present themselves. For example,
# using stress instead of stressrace and passing the '-p' stressflag which
# controls concurrency.
./scripts/gceworker.sh start && ./scripts/gceworker.sh mosh
cd ~/go/src/github.com/cockroachdb/cockroach && \
stdbuf -oL -eL \
make stressrace TESTS=TestChangefeedDescription PKG=github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl TESTTIMEOUT=5m STRESSFLAGS='-maxtime 20m -timeout 10m' 2>&1 | tee /tmp/stress.log
```

Failed test: https://teamcity.cockroachdb.com/viewLog.html?buildId=1208452&tab=buildLog

```
      /go/src/github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl/changefeed_dist.go:198 +0x1313
  github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl.(*changefeedResumer).Resume()
      /go/src/github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl/changefeed_stmt.go:449 +0x5be
  github.com/cockroachdb/cockroach/pkg/jobs.(*Registry).resume.func1()
      /go/src/github.com/cockroachdb/cockroach/pkg/jobs/registry.go:543 +0x3ff
  github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTask.func1()
      /go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:325 +0xf3
==================
I190329 07:30:06.455642 36407 util/stop/stopper.go:546  quiescing; tasks left:
15     [async] storage.IntentResolver: cleanup txn records
1      [async] transport racer
1      [async] job-438351319352934401
1      [async] intent_resolver_ir_batcher
1      [async] intent_resolver_gc_batcher
1      [async] closedts-rangefeed-subscriber
1      [async] changefeed-poller
W190329 07:30:06.455647 36712 storage/intentresolver/intent_resolver.go:822  failed to gc transaction record: could not GC completed transaction anchored at /Table/SystemConfigSpan/Start: node unavailable; try another peer
I190329 07:30:06.455713 36672 kv/transport_race.go:113  transport race promotion: ran 50 iterations on up to 345 requests
I190329 07:30:06.457580 36407 util/stop/stopper.go:546  quiescing; tasks left:
14     [async] storage.IntentResolver: cleanup txn records
1      [async] transport racer
1      [async] job-438351319352934401
1      [async] closedts-rangefeed-subscriber
1      [async] changefeed-poller
I190329 07:30:06.458020 36828 ccl/changefeedccl/changefeed_stmt.go:473  [n1] CHANGEFEED job 438351319352934401 returning with error: open /tmp/TestChangefeedDescription_poller258021152/ca: no such file or directory
I190329 07:30:06.460394 36407 util/stop/stopper.go:546  quiescing; tasks left:
14     [async] storage.IntentResolver: cleanup txn records
1      [async] transport racer
1      [async] job-438351319352934401
1      [async] closedts-rangefeed-subscriber
I190329 07:30:06.460597 36407 util/stop/stopper.go:546  quiescing; tasks left:
13     [async] storage.IntentResolver: cleanup txn records
1      [async] transport racer
1      [async] job-438351319352934401
1      [async] closedts-rangefeed-subscriber
I190329 07:30:06.460748 36407 util/stop/stopper.go:546  quiescing; tasks left:
12     [async] storage.IntentResolver: cleanup txn records
1      [async] transport racer
1      [async] job-438351319352934401
1      [async] closedts-rangefeed-subscriber
I190329 07:30:06.460869 36407 util/stop/stopper.go:546  quiescing; tasks left:
11     [async] storage.IntentResolver: cleanup txn records
1      [async] transport racer
1      [async] job-438351319352934401
1      [async] closedts-rangefeed-subscriber
I190329 07:30:06.461064 36407 util/stop/stopper.go:546  quiescing; tasks left:
9      [async] storage.IntentResolver: cleanup txn records
1      [async] transport racer
1      [async] job-438351319352934401
1      [async] closedts-rangefeed-subscriber
I190329 07:30:06.461220 36407 util/stop/stopper.go:546  quiescing; tasks left:
9      [async] storage.IntentResolver: cleanup txn records
1      [async] transport racer
1      [async] closedts-rangefeed-subscriber
I190329 07:30:06.461470 36407 util/stop/stopper.go:546  quiescing; tasks left:
4      [async] storage.IntentResolver: cleanup txn records
1      [async] transport racer
1      [async] closedts-rangefeed-subscriber
I190329 07:30:06.461623 36407 util/stop/stopper.go:546  quiescing; tasks left:
2      [async] storage.IntentResolver: cleanup txn records
1      [async] transport racer
1      [async] closedts-rangefeed-subscriber
I190329 07:30:06.461726 36407 util/stop/stopper.go:546  quiescing; tasks left:
1      [async] transport racer
1      [async] storage.IntentResolver: cleanup txn records
1      [async] closedts-rangefeed-subscriber
    --- FAIL: TestChangefeedDescription/poller (121.41s)
        testing.go:771: race detected during execution of test
    testing.go:771: race detected during execution of test

```

--- a/pkg/ccl/changefeedccl/changefeed_processors.go
+++ b/pkg/ccl/changefeedccl/changefeed_processors.go
@@ -184,8 +184,8 @@ func (ca *changeAggregator) Start(ctx context.Context) context.Context {
 	// but only the first one is ever used.
 	ca.errCh = make(chan error, 2)
 
+	ca.pollerDoneCh = make(chan struct{})
 	if err := ca.flowCtx.Stopper().RunAsyncTask(ctx, "changefeed-poller", func(ctx context.Context) {
-		ca.pollerDoneCh = make(chan struct{})
 		defer close(ca.pollerDoneCh)
 		var err error
 		if PushEnabled.Get(&ca.flowCtx.Settings.SV) {
@@ -199,6 +199,10 @@ func (ca *changeAggregator) Start(ctx context.Context) context.Context {
 		ca.errCh <- err
 		ca.cancel()
 	}); err != nil {
+		// If err != nil then the RunAsyncTask closure never ran, which means we
+		// need to manually close ca.pollerDoneCh so `(*changeAggregator).close`
+		// doesn't hang.
+		close(ca.pollerDoneCh)
 		ca.errCh <- err
 		ca.cancel()
 	}

